### PR TITLE
Enable shift+select in mouse mode

### PR DESCRIPTION
**Pull request by chrisduerr**
When an application takes control over the mouse, it usually disables
selection completely. However the common way to still make selection
possible is by allowing selection while the shift key is held down.

This feature is implemented here by making use of the new `modifiers`
field on mouse events with glutin/winit.

This fixes jwilm/alacritty#146.

I've thought about changing the cursor, but it seems like that would be kinda odd (you wouldn't want it changing every time when hitting shift). And other terminal emulators also do not seem to change the cursor with shift selection.
For clearing a selection you also need to hold shift and click, but this also seems consistent with my experience on other terminal emulators.
If you have any feedback on these two things, feel free to tell me about it.

### CODE REVIEW

**Comment by maximbaz** [comment]
Thanks, works nicely. I find it a bit odd that I have to hold Shift when clicking to clear the selection, even if other terminals do the same - I can't imagine a scenario where I would shift-select something, then start clicking without shift and expecting the selection to stay.

**Comment by chrisduerr** [comment]
Yeah I have definitely thought about just setting it up so clicking always clears selection no matter if shift is held or not. However until now alacritty also hasn't cleared when the mouse modes were enabled so I decided to stick with this for now. If anyone has a reason why the selection shouldn't be cleared when mouse mode is enabled and left mouse is clicked, please tell me.

Otherwise I'd be inclined to change it to always clear on left click. Maybe @jwilm knows more about this kind of stuff?

**Comment by unphased** [comment]
I have no preference and have found myself eventually mashing Esc to clear it when I couldn't figure it out. I could say that the terms I use (mostly iterm) do clear it on the next regular click...

One thing I wanted to bring up is that I get really good use out of Shift+Alt+drag for selecting a block. 

This is very useful in vim/tmux to omit line numbers and other vertically split panes while copying code or log output, for example.

**Comment by jwilm** [comment]
How does this compare with @gilbertw1's implementation?

**Comment by chrisduerr** [comment]
I did not see an open PR from him. But looking at his master it looks like the main difference is that he's passing only the shift mod and I'm passing all modifiers. And mine is not using the git version of glutin.

If you want to wait from a PR of him and merge that, feel free. I don't really care as long as this feature is merged. There are probably a few things that could be improved in both implementations, but they look almost identical.

**Comment by gilbertw1** [comment]
I hadn't opened a PR for it yet. As @chrisduerr said this implementation does look nearly identical to my fork with the exception that he's passing through all of the modifiers, which I believe is probably a better choice.

Looks good to me :+1:

**Comment by chrisduerr** [comment]
The single-click behavior seems to be the same for our versions from what I can tell. I've updated this PR to use only the shift key.

Now they seem to be pretty much identical, except for the test which @gilbertw1 didn't change.

**Comment by maximbaz** [comment]
So is it settled that only Shift+Click clears the selection, and not the Click alone? Still find that odd 😛

**Comment by chrisduerr** [comment]
Based on the current implementation that would be the "safe" choice. But I still am open to change this. Maybe @jwilm has an opinion on that?

I don't think it's a big deal either way, but I'd love to get some more input on this.

**Comment by DSpeckhals** [comment]
I prefer a simple click clearing the selection, but I'm not completely opposed to shift-click. This is a great feature overall!

**Comment by chrisduerr** [comment]
It seems like the common consensus is that even in mouse mode, the
selection should be cleared with a normal click.

If there is any reason why this should not be the case, please let me
know.

**Comment by unphased** [comment]
I noticed that doubleclicking with this makes the selection work word-wise. Very good!

**Comment by chrisduerr** [comment]
Double-clicking in general is used for semantic selection. And triple-clicking is used for line-selection. This PR didn't introduce this but the functionality should work when holding shift in mouse-mode now.

**Comment by unphased** [comment]
Ah yes I remember it working from before, too. Being in tmux most of the time I had forgotten.

**Comment by unphased** [comment]
BTW, I am used to how selecting copies what was selected to the system buffer right away. I wonder if a configuration can be added to make this happen?

**Comment by chrisduerr** [comment]
This should be the case for both normal selection and selection with shift when mouse-mode is enabled. Is this not working for you?

**Comment by maximbaz** [comment]
It is implemented, just keep in mind that mouse selection puts contents to the "primary selection clipboard" (something you would paste in vim with `"*p`), while pressing <kbd>Ctrl+Shift+C</kbd> on a selection puts contents to the "system clipboard" (something you would paste in vim with `"+p`). Maybe this was the confusion? 🙂

**Comment by chrisduerr** [comment]
I think he's talking about MacOS, not linux. I've talked to jwilm on IRC and he said `there's no equivalent functionality in other macOS terminals`, so directly copying to clipboard is not implemented in MacOS.

If there is a convention how this should be done (I think @unphased has mentioned that iTerm2 supports this), then implementing this feature for MacOS too should be possible.

File: src/input.rs [hunk 1]
```diff
@@ -701,8 +702,8 @@ mod tests {
                     mouse_bindings: &config.mouse_bindings()[..],
                 };
 
-                if let Event::WindowEvent { event: WindowEvent::MouseInput { state, button, .. }, .. } = $input {
-                    processor.mouse_input(state, button);
+                if let Event::WindowEvent { event: WindowEvent::MouseInput { state, button, modifiers, .. }, .. } = $input {
+                    processor.mouse_input(state, button, modifiers.shift);
```
Comment (jwilm) [inline src/input.rs]: It seems to me that `modifiers` should be passed through in its entirety. If we ever want to add some other mouse + modifier functionality, then all the mods are already being passed without bloating the method signature.

**Comment by unphased** [comment]
Thanks for the clarifications guys!

**Comment by lf94** [comment]
Am I supposed to be able to hold shift + mouse drag to select text in tmux now? Because it isn't working after a `git pull` and `cargo build --release && cargo install` :(

**Comment by chrisduerr** [comment]
@lf94 Are you sure you are on the latest master? Usually in tmux you should be able to just select text by default as long as mouse mode is not enabled. After enabling mouse mode, you should now be able to select text with shift + mouse.

At least this is the way it works on my X11 machine. Are you on macos?

**Comment by lf94** [comment]
7b4ba80bb is the first bit of the latest commit hash.

I'm on Debian Stretch (9.0).

I have mouse mode enabled in tmux, as it's always been :) (`set -g mouse on`)

**Comment by chrisduerr** [comment]
I'm not quite sure what to tell you then. It seems to work just fine with tmux 2.6 on archlinux.

What exactly happens when you try to shift-select? Do you still get the mouse selection from tmux?

**Comment by lf94** [comment]
Yeah that is exactly what is happening still.

**Comment by lf94** [comment]
My configuration:

```
set -g mouse on
set-option -g status-position top
set -g default-terminal "screen-256color"

## COLORSCHEME: gruvbox dark
set-option -g status "on"

# default statusbar colors
set-option -g status-bg default #bg1
set-option -g status-fg default #fg1

# default window title colors
set-window-option -g window-status-bg default #yellow
set-window-option -g window-status-fg default #bg1

set-window-option -g window-status-activity-bg default #bg1
set-window-option -g window-status-activity-fg default #fg3

# active window title colors
set-window-option -g window-status-current-bg default
set-window-option -g window-status-current-fg default #bg1

# pane border
set-option -g pane-active-border-fg default #fg2
set-option -g pane-border-fg default #bg1

# pane number display
set-option -g display-panes-active-colour default #fg2
set-option -g display-panes-colour default #bg1


set-option -g status-attr "underscore"
set-option -g status-justify "left"
set-option -g status-left-attr "underscore"
set-option -g status-left-length "0"
set-option -g status-right-attr "underscore"
set-option -g status-right-length "0"
set-window-option -g window-status-activity-attr "none"
set-window-option -g window-status-attr "none"
set-window-option -g window-status-separator ""

set-option -g status-left ""
set-option -g status-right ""
set-window-option -g window-status-current-format "#[fg=default, bg=colour136, bold] #I #W "
set-window-option -g window-status-format "#[fg=default, bg=default] #I #W "
```

**Comment by chrisduerr** [comment]
I'm unable to reproduce even with your config.

**Comment by maximbaz** [comment]
@lf94 is it possible that you use an unusual keyboard layout maybe? For the record, I can't reproduce the issue too.

**Comment by lf94** [comment]
I'm using Dvorak, setup via Cinnamon's configuration tools.

It works in xterm/urxvt/gnome-terminal though just fine.

**Comment by lf94** [comment]
Is there a way to verify my alacritty binary? You guys have a SHA-1 we can compare?

**Comment by lf94** [comment]
0edb8ba861... is the first bit of the hash.

**Comment by maximbaz** [comment]
I wonder if this is somehow related or caused by https://github.com/jwilm/alacritty/issues/61, specifically:

> There's an issue supporting alternative keyboard layouts with Glutin's input system...

To double check, try QWERTY maybe?

My binary, compiled on Arch Linux: [alacritty.zip](https://github.com/jwilm/alacritty/files/1602062/alacritty.zip)

**Comment by chrisduerr** [comment]
If `cat src/input.rs | grep modifiers.shift` gives you any results, you shoud definitely have the feature.
Also try executing it with `cargo run (--release)` directly.

Dvorak should not be an issue, I'm using it myself and it works just fine.

**Comment by lf94** [comment]
Yes, it gives me results.

I'm now waiting for it to compile with `cargo run --release`.

**Comment by lf94** [comment]
`cargo run --release` worked.

It seems `cargo install --force` didn't work properly!

Mission complete. Thank you guys :)

**Comment by chrisduerr** [comment]
No problem. :)

**Pull request merged**